src/app/app.cpp

```diff
@@ -0,0 +1,215 @@
+#include "glm/fwd.hpp"
+#include <iostream>
+
+#include <GLFW/glfw3.h>
+#include <bgfx/bgfx.h>
+#include <bgfx/platform.h>
+#include <bx/bx.h>
+#include <bx/math.h>
+
+#include <glm/glm.hpp>
+#include <glm/gtc/type_ptr.hpp>
+
+#if BX_PLATFORM_LINUX
+#define GLFW_EXPOSE_NATIVE_X11
+#elif BX_PLATFORM_WINDOWS
+#define GLFW_EXPOSE_NATIVE_WIN32
+#elif BX_PLATFORM_OSX
+#define GLFW_EXPOSE_NATIVE_COCOA
+#endif
+
+#include "input/keyboard.hpp"
+#include "input/mouse.hpp"
+#include "logging/logger/logger.hpp"
+#include "shader_loading/shader_loading.hpp"
+#include "window/window.hpp"
+
+#include "controllers/camera/flying_camera_controller.hpp"
+#include "app/app.hpp"
+
+struct PosColorVertex {
+    float m_x;
+    float m_y;
+    float m_z;
+    uint32_t m_abgr;
+
+    static void init() {
+        ms_decl.begin()
+                .add(bgfx::Attrib::Position, 3, bgfx::AttribType::Float)
+                .add(bgfx::Attrib::Color0, 4, bgfx::AttribType::Uint8, true)
+                .end();
+    };
+
+    static bgfx::VertexLayout ms_decl;
+};
+
+bgfx::VertexLayout PosColorVertex::ms_decl;
+
+static PosColorVertex s_cubeVertices[] = {
+        {-1.0f, 1.0f, 1.0f, 0xff000000},   {1.0f, 1.0f, 1.0f, 0xff0000ff},
+        {-1.0f, -1.0f, 1.0f, 0xff00ff00},  {1.0f, -1.0f, 1.0f, 0xff00ffff},
+        {-1.0f, 1.0f, -1.0f, 0xffff0000},  {1.0f, 1.0f, -1.0f, 0xffff00ff},
+        {-1.0f, -1.0f, -1.0f, 0xffffff00}, {1.0f, -1.0f, -1.0f, 0xffffffff},
+};
+
+static const uint16_t s_cubeTriList[] = {
+        1, 0, 2, 3, 1, 2, 6, 4, 5, 6, 5, 7, 2, 0, 4, 2, 4, 6,
+        5, 1, 3, 7, 5, 3, 4, 0, 1, 5, 4, 1, 3, 2, 6, 3, 6, 7,
+};
+
+bgfx::VertexBufferHandle m_vbh;
+bgfx::IndexBufferHandle m_ibh;
+bgfx::ProgramHandle m_program;
+
+bool s_showStats = false;
+
+App::App(const char *title, uint32_t width, uint32_t height): m_window(title, width, height) {}
+
+App::~App() {}
+
+void App::init() {
+
+    // Logger init
+    Logger::instance().init();
+    Logger::setDebugMode(true);
+
+    m_window.init();
+
+    // Call bgfx::renderFrame before bgfx::init to signal to bgfx not to
+    // create a render thread. Most graphics APIs must be used on the same
+    // thread that created the window.
+    bgfx::renderFrame();
+    // Initialize bgfx using the native window handle and window resolution.
+    bgfx::Init init;
+
+    init.platformData.nwh = m_window.getWindow();
+
+    int width = 1024, height = 768;
+    m_window.getSize(&width, &height);
+    init.resolution.width = (uint32_t)width;
+    init.resolution.height = (uint32_t)height;
+    init.resolution.reset = BGFX_RESET_VSYNC;
+    init.type = bgfx::RendererType::OpenGL;
```

Comment:
Just a tip for the future, it might be useful to create separate private methods like initWindow() initRenderer(), etc because the amount of code may drastically increase => will be hard to read